Do not show empty tags for digest references in output

dmcgowan opened the pull request:
When a repository has a tag and digests, show tag for each digest value. Rows will not be duplicated for the same image name which has both a tag and a digest.

Ping @aaronlehmann @endophage @dnephin 

```
$ DOCKER_CONTENT_TRUST=1 docker pull ubuntu                           
Using default tag: latest
Pull (1 of 1): ubuntu:latest@sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4
sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4: Pulling from library/ubuntu
5ba4f30e5bea: Pull complete 
9d7d19c9dc56: Pull complete 
ac6ad7efd0f9: Pull complete 
e7491a747824: Pull complete 
a3ed95caeb02: Pull complete 
Digest: sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4
Status: Downloaded newer image for ubuntu@sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4
Tagging ubuntu@sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4 as ubuntu:latest
$ docker images
REPOSITORY          TAG                          IMAGE ID            CREATED              SIZE
ubuntu              latest                       2fa927b5cdd3        2 weeks ago          122 MB
$ docker images --digests
REPOSITORY          TAG                          DIGEST                                                                    IMAGE ID            CREATED              SIZE
ubuntu              latest                       sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4   2fa927b5cdd3        2 weeks ago          122 MB
```

dmcgowan commented:
Solution for #23503 and allowing content trust by default

dmcgowan commented:
Updated after discussing with @aaronlehmann. Since the multiple digest case for the same image id is considered rare, updated the logic to show the each digest for each tag of the same repo name. This will normally result in a single row output when `docker images --digests` since different digests usually result in different image ids.

dmcgowan commented:
Illustrated different pulling image with 2 different tags. With schema1 manifests 2 different manifests will be created causing the case of multiple digests showing up. Since manifest tags are not enforced, the corresponding images may be pulled by either manifest hash. However the common case of issuing `docker images` without the digest flags will still show the shorter list.

```
$ docker images          
REPOSITORY          TAG                          IMAGE ID            CREATED             SIZE
alpine              3.4                          f70c828098f5        12 days ago         4.799 MB
ubuntu              16.04                        2fa927b5cdd3        2 weeks ago         122 MB
ubuntu              latest                       2fa927b5cdd3        2 weeks ago         122 MB
ubuntu              15.10                        0b4516a442e7        2 weeks ago         136.8 MB
$ docker images --digests
REPOSITORY          TAG                          DIGEST                                                                    IMAGE ID            CREATED             SIZE
alpine              3.4                          sha256:7d7d68e60bec044b1821208884a2a4c9de64dfd5807f04cd32ffa5f5232d7848   f70c828098f5        12 days ago         4.799 MB
ubuntu              16.04                        sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4   2fa927b5cdd3        2 weeks ago         122 MB
ubuntu              16.04                        sha256:f5edf3b741a08b573eca6bf25257847613540538a17b86e2b76e14724a0be68a   2fa927b5cdd3        2 weeks ago         122 MB
ubuntu              latest                       sha256:46fb5d001b88ad904c5c732b086b596b92cfb4a4840a3abd0e35dbb6870585e4   2fa927b5cdd3        2 weeks ago         122 MB
ubuntu              latest                       sha256:f5edf3b741a08b573eca6bf25257847613540538a17b86e2b76e14724a0be68a   2fa927b5cdd3        2 weeks ago         122 MB
ubuntu              15.10                        sha256:d010a2154cfbead5674bcadd358003f05a5935337b176cb032ec9e6a36369e70   0b4516a442e7        2 weeks ago         136.8 MB
```

aaronlehmann commented on api/client/formatter/formatter.go:
```diff
@@ -200,42 +204,98 @@ virtual_size: {{.Size}}
+			for repo, tags := range repoTags {
+				digests := repoDigests[repo]
+
+				// Only display digests with tags
```
I think the comment is slightly misleading. This is displaying digests with tags for repos that have at least one tag associated with them.

aaronlehmann commented:
LGTM

stevvooe commented:
LGTM

dmcgowan committed:
Do not show empty tags for digest references in output

When a repository has a tag and digests, show tag for each digest value.
Do not duplicate rows for the same image name with both tag and digest.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dnephin commented:
Thanks for this PR. I think the builds timed out, so I've restarted them.

I'm going to cherry-pick this commit into my PR as well to get the tests passing.

aaronlehmann commented:
Added to 1.12.0 milestone since this is blocking #23503.

icecrime commented:
Excellent, thanks a lot! LGTM.
